### PR TITLE
Track from mesh

### PR DESCRIPTION
**Pull request by kristianeschenburg**
Added two methods to dipy/tracking/utils.py:

``seeds_from_mesh``: generates seed points randomly sampled from a surface mesh.  Parameterized by vertex coordinates, list of triangles in mesh, number of samples to generate per face, and an affine transformation.  The affine transformation brings the generated seed points into the tracking space using the inverse of the diffusion image affine transformation.  Assumes that the mesh is already aligned to some standard space, and that the diffusion image affine matrix brings the diffusion image to this standard space. 

``sample_triangle``: samples points from a single mesh triangle, defined by 3 vertex coordinates.

I've included a description of each method, with the types of accepted parameters, return values, exceptions raised, as well as examples, though I still need to write tests.

A while back, @MrBago had suggested (https://mail.python.org/pipermail/neuroimaging/2016-August/001128.html) also projecting vertices a few **mm** into the white matter -- I'm hoping to add a method to do this as well.  Ideally, this could be extended to track *between* surfaces, though there are a few issues that would need to be resolved, including:
1. stopping criteria for tracking (streamline / face intersection)
2. how to update counts for source/target points upon streamline intersection

Any thoughts, critiques, comments, and ideas are much appreciated!

### CODE REVIEW

**Comment by Garyfallidis** [comment]
Thank you @kristianeschenburg @StongeEtienne can you have a look at this PR? Seem relevant to your work.

**Comment by StongeEtienne** [comment]
I have very similar code currently in my branch of Scilpy, I could share it.

I would not mind to do a full code review; 
e.g. the "for" loop is not necessary and could be optimized with numpy operation.

I don't want to remove credit from the author, maybe we could arrange a meeting for 
"surface integration in dipy", I have alot of tools in our internal Scilpy (script and functions) 
that could be integrated in dipy. ex: a mapping from endpoints to surface vertices, streamlines intersection with surface (using VTK), my surface flow from https://github.com/StongeEtienne/trimeshpy.

@gabknight  is also currently integration initial direction, https://github.com/nipy/dipy/pull/1627
made to interface with surface seeding + directional orientation (the surface inner normal, in my case).

**Comment by kristianeschenburg** [comment]
@Garyfallidis @StongeEtienne thank you for the feedback!  I'd appreciate the code review (always looking to improve my development skills). I'm interested in having the "surface integration in dipy" meeting, should we schedule a meeting to discuss further?

Likewise, after having looked at your paper [here](https://www.ncbi.nlm.nih.gov/pubmed/29258891) @StongeEtienne, I think that would make an excellent contribution -- I'm also keen on applying your surfaceflow and Scilpy functionality to my own work.

**Comment by StongeEtienne** [comment]
@kristianeschenburg Sorry for the delay.

After discussing with Maxime Descoteaux, and @jchoude, 
we should try to schedule meeting next week, for the **integration of surfaces in Dipy**.
I think @Garyfallidis might be busy (for obvious reason, congratulation !), but we could start thinking about this with @skoudoro @arokem @gabknight .

Best,

**Comment by Garyfallidis** [comment]
@StongeEtienne Wednesday morning 9 am EST is good with me. Let me know if that works with you.

**Comment by StongeEtienne** [comment]
@Garyfallidis 
For me it's possible, and might be for @jchoude and @gabknight,
but it might not be for @kristianeschenburg on the west coast.

**Comment by kristianeschenburg** [comment]
@StongeEtienne @Garyfallidis 

Yes, sorry for the delay, I took a few days off before the start of the new quarter.  9am EST would be OK with me if this is most convenient for everyone, though later in the morning would be better.  Is everyone free tomorrow or Friday morning?

File: dipy/tracking/utils.py [hunk 1]
```diff
@@ -412,6 +412,145 @@ def seeds_from_mask(mask, density=[1, 1, 1], voxel_size=None, affine=None):
...
+
+        for c, face in enumerate(faces):
+
+            # Get coordinates of each vertex in triangle
+            coords = vertices[face, :]
+            # Generate seeds from triangle and compute nearest vertex
+            face_seeds, nearest = sample_triangle(coords, face, n_samples)
+
+            lo = c*n_samples
+            hi = (c+1)*n_samples
```
Comment (nilgoyette) [inline dipy/tracking/utils.py]: [PEP8](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements). I'm not sure about DiPy rules, but there's almost always spaces around operators.

File: dipy/tracking/utils.py [hunk 2]
```diff
@@ -412,6 +412,145 @@ def seeds_from_mask(mask, density=[1, 1, 1], voxel_size=None, affine=None):
...
+
+            seeds[lo:hi, :] = face_seeds
+            assignment[lo:hi] = nearest
+
+    if affine is not None:
+        # Use affine to move seeds into real world coordinates
+        seeds += affine[:3, 3]
+        seeds = np.dot(seeds, affine[:3, :3].T)
+
+    return [seeds, assignment]
```
Comment (nilgoyette) [inline dipy/tracking/utils.py]: Those elements are unrelated to each other, they do not belong in a list. Why not

    return seeds, assignment

File: dipy/tracking/utils.py [hunk 3]
```diff
@@ -412,6 +412,145 @@ def seeds_from_mask(mask, density=[1, 1, 1], voxel_size=None, affine=None):
...
+    if n_samples is None:
+        return vertices
+    # Otherwise, check n_samples type and value
+    else:
+        if not isinstance(n_samples, int):
+            raise TypeError('n_samples must be of type int')
+        if n_samples < 0:
+            raise ValueError('n_samples must be greater than 0')
+
+    v1,v2,v3 = vertices[0:3,:]
```
Comment (nilgoyette) [inline dipy/tracking/utils.py]: PEP8 also in this function. `v1, v2, v3 = ...`, spaces around operators, etc.

File: dipy/tracking/utils.py [hunk 4]
```diff
@@ -412,6 +412,145 @@ def seeds_from_mask(mask, density=[1, 1, 1], voxel_size=None, affine=None):
...
+        when ``n_samples`` not of type int
+    ValueError
+        when ``n_samples`` less than 0
+    """
+
+    # If n_samples not provided, return original vertex coordinates
+    if n_samples is None:
+        return vertices
+    # Otherwise, check n_samples type and value
+    else:
```
Comment (nilgoyette) [inline dipy/tracking/utils.py]: Arguable, but `else` is useless after a `return`.

File: dipy/tracking/utils.py [hunk 5]
```diff
@@ -412,6 +412,145 @@ def seeds_from_mask(mask, density=[1, 1, 1], voxel_size=None, affine=None):
...
+    # generate barycentric average of vertex coordinates
+    samples = (1-np.sqrt(r1))*v1 + np.sqrt(r1)*(1-r2)*v2 + \
+        np.sqrt(r1)*r2*v3
+    
+    # Compute distance to each vertex in triangle
+    # and assign seeds to nearest vertex
+    distance = cdist(vertices, samples)
+    nearest = np.asarray(face)[np.argmin(distance, axis=0)]
+
+    return [samples, nearest]
```
Comment (nilgoyette) [inline dipy/tracking/utils.py]: Idem, simply return the objects without a list or tuple or anything.

**Comment by skoudoro** [comment]
Hi @kristianeschenburg,

I hope you doing well! What is the status of this work?  Can you look at #1648, it seems related to your work here and it has been merged. We have now a new `mesh` module that can help you.

Let us know what you plan to do and if I should close this PR, etc.... 

Also @StongeEtienne, could you look at this PR?

Thank you both for the feedback.

**Comment by skoudoro** [comment]
No news concerning this PR and I believe that #1648 contains most of it. 

I am closing it for now but feel free to reopen it!
